[SPARK-26985][CORE] Fix "access only some column of the all of columns " for big endian architecture

ketank-new opened the pull request:
## What changes were proposed in this pull request?

Changes are related to BIG ENDIAN system
This changes are done to 
1. identify s390x platform.
2. use byteorder to BIG_ENDIAN for big endian systems
3. changes for 2 are done in access functions putFloats() and putDouble()



## How was this patch tested?

Changes have been tested to build successfully on s390x as well x86 platform to make sure build is successful.

ketank-new committed:
s390x specific changes

ketank-new committed:
s390x specific changes

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

kiszk commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java:
```diff
@@ -396,7 +396,7 @@ public void putFloats(int rowId, int count, byte[] src, int srcIndex) {
           Platform.DOUBLE_ARRAY_OFFSET + rowId * 4L, count * 4L);
     } else {
-      ByteBuffer bb = ByteBuffer.wrap(src).order(ByteOrder.LITTLE_ENDIAN);
+      ByteBuffer bb = ByteBuffer.wrap(src).order(ByteOrder.BIG_ENDIAN);
```
Could you please let us know which test cases are failed due to this code on `s390x`?

gatorsmile replied:
This change looks risky. Any reason?

kiszk replied:
I agree. Once we know the test cases that cause failures, I will double-check by running them on another big endian environment.

ketank-new replied:
below are the test suites which had some test failing initially and then passed on doing the above source code changes

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/compression/PassThroughEncodingSuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/DataFrameTungstenSuite.scala

srowen replied:
Hm, I guess the changes make sense? we don't test the big-endian branch on Jenkins, so I'm not surprised it still passes Jenkins. For a big-endian platform, indeed, wouldn't you have to read floats from the ByteBuffer as big-endian?

kiszk replied:
Thanks, I will test them on ppc64 linux (BE).

kiszk commented:
Could you please add JIRA number `[SPARK-....]` and `[core]` as a prefix of the title?

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -51,6 +51,8 @@
       // Since java.nio.Bits.unaligned() doesn't return true on ppc (See JDK-8165231), but
       // ppc64 and ppc64le support it
       _unaligned = true;
+    } else if (arch.matches("s390x")) {
```
`.contains`? that isn't a regex. You can put it into the clause above and modify the comment

srowen replied:
This can still be changed

ketank-new replied:
@srowen  : did not get your <.contains> comment what exactly you are trying to say
I have intentionally written a else if blk to keep processor specific changes in seperate block.

are you asking me to add "s390x" in the same "ppc64le" blk ... let me know

srowen replied:
Yes, keep them together. And it does not need to be a regex with `matches`; just use `contains`.

ketank-new replied:
> @ketank-new Could you address the comments and resolve the conflicts?

sure

srowen commented:
This needs a better title -- 'changes' isn't descriptive. What problem are you solving? It will need a JIRA if the change is more than simply including it as a big-endian platform, as the rest is pretty non-trivial

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented:
This seems to fail in Jenkins, doesn't it?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11465/
Test PASSed.

SparkQA commented:
**[Test build #106217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106217/testReport)** for PR 24788 at commit [`fdbb722`](https://github.com/apache/spark/commit/fdbb722bdb61881728be8e8dcb6bce5f5648048e).

SparkQA commented:
**[Test build #106217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106217/testReport)** for PR 24788 at commit [`fdbb722`](https://github.com/apache/spark/commit/fdbb722bdb61881728be8e8dcb6bce5f5648048e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106217/
Test PASSed.

ketank-new commented:
below are the test suites which had some test failing initially and then passed on doing the above source code changes

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/compression/PassThroughEncodingSuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala

spark/sql/core/src/test/scala/org/apache/spark/sql/DataFrameTungstenSuite.scala

kiszk commented:
In a big-endian testing environment, if necessary, I could contact the usage of [this envionment](http://osuosl.org/services/powerdev/request_powerci/)

HyukjinKwon commented:
@ketank-new, please file a JIRA with error message with problem analysis and describe how this PR fixes in PR description. Otherwise, no one knows what's going on about what you faced.

ketank-new commented:
@HyukjinKwon : I do not mind raising a new JIRA for the above changes
But let me inform you that we have been continuously in discussion on JIRA regarding this earlier too
Please refer SPARK-26985

HyukjinKwon commented:
If that's the exact same problem, please link that JIRA to the PR title (see https://spark.apache.org/contributing.html).

Also, please clarify why changing from little endian to big endian is safe in little endian OSes.

ketank-new commented:
> Could you please add JIRA number `[SPARK-....]` and `[core]` as a prefix of the title?

done

ketank-new commented:
> If that's the exact same problem, please link that JIRA to the PR title (see https://spark.apache.org/contributing.html).

Linked by changing the Title above.. were in the JIRA number is seen
> 
> Also, please clarify why changing from little endian to big endian is safe in little endian OSes.

Clarification for the changes
putFloats() and putDoubles() from files OffHeapColumnVector.java and OnHeapColumnVector.java
do get called whenever test cases use float and double data respectively.

If you check the definitions of putFloat() and putDouble() for a BIG_ENDIAN system, the control moves into the else block,within the else block the byteorder which is set is LITTLE_ENDIAN which is exactly opposite to what is expected on a BIG_ENDIAN system

changing this byteorder to BIG_ENDIAN represents float and double values as expected on a BIG_ENDIAN processor and hence further helps in passing the written test cases.

With this changes done and running the test cases on LITTLE_ENDIAN moves the control to if block and thereby also passes test cases for LITTLE_ENDIAN system.

Thereby i conclude that the changes are working and test for both types of systems that is LITTLE_ENDIAN and BIG_ENDIAN

viirya commented:
@ketank-new Can you update the PR description like above clarification of changes? It can explain what this PR wants to fix.

ketank-new commented:
> @ketank-new Can you update the PR description like above clarification of changes? It can explain what this PR wants to fix.

done

srowen reviewed (changes requested):
I'll fix the title for you; PRs need descriptive title

kiszk commented:
I double-checked some of tests. I confirmed this PR fixes existing failures on big-endian platform. Thank you very much.

**Before**
```
$ uname -a
Linux ppc64be.novalocal 4.5.7-300.fc24.ppc64 #1 SMP Fri Jun 10 20:29:32 UTC 2016 ppc64 ppc64 ppc64 GNU/Linux
$ build/sbt "sql/testOnly  *.ColumnVectorSuite"
...
[info] ColumnVectorSuite:
[info] - boolean (194 milliseconds)
[info] - byte (4 milliseconds)
[info] - short (4 milliseconds)
[info] - int (4 milliseconds)
[info] - date (4 milliseconds)
[info] - long (4 milliseconds)
[info] - timestamp (5 milliseconds)
[info] - float (4 milliseconds)
[info] - double (4 milliseconds)
[info] - string (14 milliseconds)
[info] - binary (8 milliseconds)
[info] - mutable ColumnarRow (4 milliseconds)
[info] - array (1 millisecond)
[info] - struct (1 millisecond)
[info] - [SPARK-22092] off-heap column vector reallocation corrupts array data (3 milliseconds)
[info] - [SPARK-22092] off-heap column vector reallocation corrupts struct nullability (3 milliseconds)
[info] - CachedBatch boolean Apis (46 milliseconds)
[info] - CachedBatch byte Apis (7 milliseconds)
[info] - CachedBatch short Apis (8 milliseconds)
[info] - CachedBatch int Apis (9 milliseconds)
[info] - CachedBatch long Apis (8 milliseconds)
[info] - CachedBatch float Apis *** FAILED *** (30 milliseconds)
[info]   4.6006E-41 did not equal 1.0 (ColumnVectorSuite.scala:424)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:528)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:527)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:501)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$71(ColumnVectorSuite.scala:424)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$71$adapted(ColumnVectorSuite.scala:422)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:158)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$70(ColumnVectorSuite.scala:422)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$70$adapted(ColumnVectorSuite.scala:417)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.withVector(ColumnVectorSuite.scala:34)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.withVectors(ColumnVectorSuite.scala:41)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$68(ColumnVectorSuite.scala:417)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:149)
[info]   at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
[info]   at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:56)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
[info]   at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:56)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
[info]   at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
[info]   at org.scalatest.Suite.run(Suite.scala:1147)
[info]   at org.scalatest.Suite.run$(Suite.scala:1129)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
[info]   at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
[info]   at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:56)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:56)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:296)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:286)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:748)
[info] - CachedBatch double Apis *** FAILED *** (11 milliseconds)
[info]   3.03865E-319 did not equal 1.0 (ColumnVectorSuite.scala:448)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:528)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:527)
[info]   at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:501)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$75(ColumnVectorSuite.scala:448)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$75$adapted(ColumnVectorSuite.scala:446)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:158)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$74(ColumnVectorSuite.scala:446)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$74$adapted(ColumnVectorSuite.scala:441)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.withVector(ColumnVectorSuite.scala:34)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.withVectors(ColumnVectorSuite.scala:41)
[info]   at org.apache.spark.sql.execution.vectorized.ColumnVectorSuite.$anonfun$new$72(ColumnVectorSuite.scala:441)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:149)
[info]   at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
[info]   at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:56)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
[info]   at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:56)
[info]   at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
[info]   at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
[info]   at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
[info]   at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
[info]   at org.scalatest.Suite.run(Suite.scala:1147)
[info]   at org.scalatest.Suite.run$(Suite.scala:1129)
[info]   at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
[info]   at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
[info]   at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:56)
[info]   at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:56)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:507)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:296)
[info]   at sbt.ForkMain$Run$2.call(ForkMain.java:286)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:748)
[info] ScalaTest
[info] Run completed in 6 seconds, 911 milliseconds.
[info] Total number of tests run: 23
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 21, failed 2, canceled 0, ignored 0, pending 0
[info] *** 2 TESTS FAILED ***
[error] Failed: Total 23, Failed 2, Errors 0, Passed 21
[error] Failed tests:
[error] 	org.apache.spark.sql.execution.vectorized.ColumnVectorSuite
[error] (sql/test:testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 700 s, completed Jun 9, 2019 11:23:46 PM
```

**After**
```
$ uname -a
Linux ppc64be.novalocal 4.5.7-300.fc24.ppc64 #1 SMP Fri Jun 10 20:29:32 UTC 2016 ppc64 ppc64 ppc64 GNU/Linux
$ build/sbt "sql/testOnly  *.ColumnVectorSuite"
...
[info] ColumnVectorSuite:
[info] - boolean (187 milliseconds)
[info] - byte (6 milliseconds)
[info] - short (6 milliseconds)
[info] - int (5 milliseconds)
[info] - date (4 milliseconds)
[info] - long (4 milliseconds)
[info] - timestamp (4 milliseconds)
[info] - float (4 milliseconds)
[info] - double (4 milliseconds)
[info] - string (13 milliseconds)
[info] - binary (8 milliseconds)
[info] - mutable ColumnarRow (5 milliseconds)
[info] - array (1 millisecond)
[info] - struct (2 milliseconds)
[info] - [SPARK-22092] off-heap column vector reallocation corrupts array data (4 milliseconds)
[info] - [SPARK-22092] off-heap column vector reallocation corrupts struct nullability (3 milliseconds)
[info] - CachedBatch boolean Apis (45 milliseconds)
[info] - CachedBatch byte Apis (7 milliseconds)
[info] - CachedBatch short Apis (7 milliseconds)
[info] - CachedBatch int Apis (9 milliseconds)
[info] - CachedBatch long Apis (8 milliseconds)
[info] - CachedBatch float Apis (8 milliseconds)
[info] - CachedBatch double Apis (7 milliseconds)
[info] ScalaTest
[info] Run completed in 6 seconds, 733 milliseconds.
[info] Total number of tests run: 23
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 23, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 23, Failed 0, Errors 0, Passed 23
[success] Total time: 38 s, completed Jun 9, 2019 11:52:54 PM
```

dongjoon-hyun commented:
Thank you for confirming, @kiszk .

dongjoon-hyun commented:
@ketank-new Could you address the comments and resolve the conflicts?

ketank-new committed:
Merge branch 'master' into ketan_new_v2.3.2

ketank-new commented:
> @ketank-new Could you address the comments and resolve the conflicts?

done

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11624/
Test PASSed.

SparkQA commented:
**[Test build #106379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106379/testReport)** for PR 24788 at commit [`2cc30b4`](https://github.com/apache/spark/commit/2cc30b477727aa2b38648e015fea0310558f25a9).

SparkQA commented:
**[Test build #106379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106379/testReport)** for PR 24788 at commit [`2cc30b4`](https://github.com/apache/spark/commit/2cc30b477727aa2b38648e015fea0310558f25a9).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106379/
Test FAILed.

kiszk commented:
retest this please

LantaoJin committed:
[SPARK-27899][SQL] Make HiveMetastoreClient.getTableObjectsByName available in ExternalCatalog/SessionCatalog API

## What changes were proposed in this pull request?

The new Spark ThriftServer SparkGetTablesOperation implemented in https://github.com/apache/spark/pull/22794 does a catalog.getTableMetadata request for every table. This can get very slow for large schemas (~50ms per table with an external Hive metastore).
Hive ThriftServer GetTablesOperation uses HiveMetastoreClient.getTableObjectsByName to get table information in bulk, but we don't expose that through our APIs that go through Hive -> HiveClientImpl (HiveClient) -> HiveExternalCatalog (ExternalCatalog) -> SessionCatalog.

If we added and exposed getTableObjectsByName through our catalog APIs, we could resolve that performance problem in SparkGetTablesOperation.

## How was this patch tested?

Add UT

Closes #24774 from LantaoJin/SPARK-27899.

Authored-by: LantaoJin <jinlantao@gmail.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11627/
Test PASSed.

SparkQA commented:
**[Test build #106382 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106382/testReport)** for PR 24788 at commit [`2cc30b4`](https://github.com/apache/spark/commit/2cc30b477727aa2b38648e015fea0310558f25a9).

SparkQA commented:
**[Test build #106382 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106382/testReport)** for PR 24788 at commit [`2cc30b4`](https://github.com/apache/spark/commit/2cc30b477727aa2b38648e015fea0310558f25a9).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106382/
Test FAILed.

viirya commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -44,6 +44,28 @@
   public static final int DOUBLE_ARRAY_OFFSET;
 
   private static final boolean unaligned;
+  static {
```
Ur, #24825 touched `unaligned` initialization and just got merged. Please rebase with the master.

ketank-new replied:
@viirya  : are u asked me to rebase with master for Platform.java changes??

viirya replied:
Yes, otherwise it can't be built.

wangyum committed:
[SPARK-27934][SQL][TEST] Port case.sql

## What changes were proposed in this pull request?

This PR is to port case.sql from PostgreSQL regression tests. https://github.com/postgres/postgres/blob/REL_12_BETA1/src/test/regress/sql/case.sql

The expected results can be found in the link: https://github.com/postgres/postgres/blob/REL_12_BETA1/src/test/regress/expected/case.out

When porting the test cases, found one PostgreSQL specific features that do not exist in Spark SQL:

- [SPARK-27930](https://issues.apache.org/jira/browse/SPARK-27930): Add built-in Math Function: RANDOM

## How was this patch tested?

N/A

Closes #24782 from wangyum/SPARK-27934.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

HyukjinKwon committed:
[SPARK-27995][PYTHON] Note the difference between str of Python 2 and 3 at Arrow optimized

## What changes were proposed in this pull request?

When Arrow optimization is enabled in Python 2.7,

```python
import pandas
pdf = pandas.DataFrame(["test1", "test2"])
df = spark.createDataFrame(pdf)
df.show()
```

I got the following output:

```
+----------------+
|               0|
+----------------+
|[74 65 73 74 31]|
|[74 65 73 74 32]|
+----------------+
```

This looks because Python's `str` and `byte` are same. it does look right:

```python
>>> str == bytes
True
>>> isinstance("a", bytes)
True
```

To cut it short:

1. Python 2 treats `str` as `bytes`.
2. PySpark added some special codes and hacks to recognizes `str` as string types.
3. PyArrow / Pandas followed Python 2 difference

To fix, we have two options:

1. Fix it to match the behaviour to PySpark's
2. Note the differences

 but Python 2 is deprecated anyway. I think it's better to just note it and for go option 2.

## How was this patch tested?

Manually tested.

Doc was checked too:

![Screen Shot 2019-06-11 at 6 40 07 PM](https://user-images.githubusercontent.com/6477701/59261402-59ad3b00-8c78-11e9-94a6-3236a2c338d4.png)

Closes #24838 from HyukjinKwon/SPARK-27995.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

yeshengm committed:
[SPARK-21136][SQL] Disallow FROM-only statements and show better warnings for Hive-style single-from statements

Current Spark SQL parser can have pretty confusing error messages when parsing an incorrect SELECT SQL statement. The proposed fix has the following effect.

BEFORE:
```
spark-sql> SELECT * FROM test WHERE x NOT NULL;
Error in query:
mismatched input 'FROM' expecting {<EOF>, 'CLUSTER', 'DISTRIBUTE', 'EXCEPT', 'GROUP', 'HAVING', 'INTERSECT', 'LATERAL', 'LIMIT', 'ORDER', 'MINUS', 'SORT', 'UNION', 'WHERE', 'WINDOW'}(line 1, pos 9)

== SQL ==
SELECT * FROM test WHERE x NOT NULL
---------^^^
```
where in fact the error message should be hinted to be near `NOT NULL`.

AFTER:
```
spark-sql> SELECT * FROM test WHERE x NOT NULL;
Error in query:
mismatched input 'NOT' expecting {<EOF>, 'AND', 'CLUSTER', 'DISTRIBUTE', 'EXCEPT', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OR', 'ORDER', 'MINUS', 'SORT', 'UNION', 'WINDOW'}(line 1, pos 27)

== SQL ==
SELECT * FROM test WHERE x NOT NULL
---------------------------^^^
```

In fact, this problem is brought by some problematic Spark SQL grammar. There are two kinds of SELECT statements that are supported by Hive (and thereby supported in SparkSQL):
* `FROM table SELECT blahblah SELECT blahblah`
* `SELECT blah FROM table`

*Reference* [HiveQL single-from stmt grammar](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g)

It is fine when these two SELECT syntaxes are supported separately. However, since we are currently supporting these two kinds of syntaxes in a single ANTLR rule, this can be problematic and therefore leading to confusing parser errors. This is because when a  SELECT clause was parsed, it can't tell whether the following FROM clause actually belongs to it or is just the beginning of a new `FROM table SELECT *` statement.

## What changes were proposed in this pull request?
1. Modify ANTLR grammar to fix the above-mentioned problem. This fix is important because the previous problematic grammar does affect a lot of real-world queries. Due to the previous problematic and messy grammar, we refactored the grammar related to `querySpecification`.
2. Modify `AstBuilder` to have separate visitors for `SELECT ... FROM ...` and `FROM ... SELECT ...` statements.
3. Drop the `FROM table` statement, which is supported by accident and is actually parsed in the wrong code path. Both Hive and Presto do not support this syntax.

## How was this patch tested?
Existing UTs and new UTs.

Closes #24809 from yeshengm/parser-refactor.

Authored-by: Yesheng Ma <kimi.ysma@gmail.com>
Signed-off-by: Xingbo Jiang <xingbo.jiang@databricks.com>

viirya committed:
[SPARK-27701][SQL] Extend NestedColumnAliasing to general nested field cases including GetArrayStructField

## What changes were proposed in this pull request?

`NestedColumnAliasing` rule covers `GetStructField` only, currently. It means that some nested field extraction expressions aren't pruned. For example, if only accessing a nested field in an array of struct (`GetArrayStructFields`), this column isn't pruned.

This patch extends the rule to cover general nested field cases, including `GetArrayStructFields`.
## How was this patch tested?

Added tests.

Closes #24599 from viirya/nested-pruning-extract-value.

Lead-authored-by: Liang-Chi Hsieh <viirya@gmail.com>
Co-authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

ketank-new committed:
Merge branch 'ketan_new_v.3.2' of https://github.com/linux-on-ibm-z/spark into ketan_new_v2.3.2

SparkQA commented:
**[Test build #106417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106417/testReport)** for PR 24788 at commit [`9aa423f`](https://github.com/apache/spark/commit/9aa423f7e1e82cfc9b3046637ae35140ece6c960).

SparkQA commented:
**[Test build #106417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106417/testReport)** for PR 24788 at commit [`9aa423f`](https://github.com/apache/spark/commit/9aa423f7e1e82cfc9b3046637ae35140ece6c960).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106417/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11661/
Test PASSed.

srowen commented on docs/sql-migration-guide-upgrade.md:
```diff
@@ -23,6 +23,7 @@ license: |
 {:toc}
 
 ## Upgrading From Spark SQL 2.4 to 3.0
+  - In Spark version 2.4 and earlier, SQL queries such as `FROM <table>` or `FROM <table> UNION ALL FROM <table>` are supported by accident. In hive-style `FROM <table> SELECT <expr>`, the `SELECT` clause is not negligible. Neither Hive nor Presto support this syntax. Therefore we will treat these queries as invalid since Spark 3.0.
```
I think you included a bunch of unrelated changes. Can you revert or open a new PR?

ketank-new replied:
yes .. looks like .. i better open a new PR

HyukjinKwon commented:
Closing per #24861

HyukjinKwon commented:
Closing per #24861

HyukjinKwon commented:
My cat reopened .. closing again.
